orc8r/cloud/go/services/certifier/tokens.go

```diff
@@ -0,0 +1,92 @@
+package certifier
+
+import (
+	"crypto/rand"
+	"errors"
+	"hash/crc32"
+	"strings"
+
+	"github.com/jxskiss/base62"
+)
+
+type TokenPrefix string
+
+const (
+	Personal TokenPrefix = "op"
+)
+
+// GenerateToken generates a random 32-byte token with a checksum
+func GenerateToken(prefix TokenPrefix) (string, error) {
+	// generate 32 random bytes
+	bytes := make([]byte, 32)
+	_, err := rand.Read(bytes)
+	if err != nil {
+		return "", err
+	}
+	// generate checksum bytes
+	checksum := crc32.ChecksumIEEE(bytes)
+	checksumBytes := i32tob(checksum)
+	// combine to form final token
+	finalBytes := append(bytes, checksumBytes...)
+	token := base62.StdEncoding.EncodeToString(finalBytes)
+
+	return string(prefix) + "_" + token, nil
+}
+
+// ValidateToken makes sure the token has the appropriate header and
+// that the checksum is correct
+func ValidateToken(token string) error {
+	value, err := stripTokenHeader(token)
+	if err != nil {
+		return err
+	}
+	if err := validateTokenChecksum(value); err != nil {
+		return err
+	}
+	return nil
+}
+
+func stripTokenHeader(token string) (string, error) {
+	s := strings.Split(token, "_")
+	prefix, value := s[0], s[1]
+
+	if len(s) != 2 {
+		return "", errors.New("missing token prefix")
+	}
+
+	// validate token prefix
+	switch prefix {
+	case string(Personal):
+		return value, nil
+	}
+	return "", errors.New("invalid token prefix")
+}
+
+func validateTokenChecksum(token string) error {
+	bytes, err := base62.DecodeString(token)
+	if err != nil {
+		return err
+	}
+	orignalChecksum := btoi32(bytes[len(bytes)-4:])
```

Comment:
1. Check the bytes is actually long enough to perform this slice
2. "4" is a magic number -- use a const with a doc comment